enable_disable_resources.md

```diff
@@ -0,0 +1,366 @@
+what changed.
+
+This quickly leads to circular dependencies (e.g., code at the beginning of the Tiltfile
+that wants to know what services will be enabled at the end of the Tiltfile).
+
+### No New Tiltfile State
+
+Per above, we want the Tiltfile API to converge towards Kubernetes API conventions.
+
+We do NOT think that adding special hooks and extension points to the Tiltfile
+execution model is the right answer.  We don't want special Tiltfile data flow or overrides.
+
+We may have tiltfile functions that are auto-generated from API objects. We may
+also change existing Tiltfile functions to add new fields to API objects that
+they assemble.
+
+## Non-Goals
+
+### User Interfaces
+
+The user interface for disabling a service is beyond the scope of this doc.
+
+### Detailed API Specifications
+
+Although this doc contains some Go structs, please don't interpret those Go
+structs as detailed API specifications!
+
+They're mainly intended to be illustative.
+
+For this doc, we're more interested in broad categories of solutions rather than
+exact field names. We expect to revise field names as we implement the feature.
+
+## Possible Approaches
+
+The approaches here aren't necessarily mutually exclusive.
+
+### DisableOnSpec / DisableOnStatus
+
+If an API object can be disabled, then that API object should have two new
+fields for processing "disable" triggers: `spec.disableOn` and
+`status.disableOn`.
+
+Here's what that might look like:
+
+```go
+type KubernetesApplySpec struct {
+  ...
+  
+  DisableOn *DisableOnSpec
+}
+
+type KubernetesApplyStatus struct {
+  ...
+  
+  DisableOn *DisableOnStatus
+}
+
+type DisableOnSpec struct {
+  UIButtons []*UIButton
+}
+
+type DisableOnStatus struct {
+  Disabled bool
+  LastUpdateTime metav1.Time
+  Reason string
+}
+```
+
+To disable a resource, the UI POSTs a status update to the UIButton. The
+KubernetesApply controller consumes that status update and toggles the
+enabled/disabled state.
+
+The UIResource controller aggregates all the DisableOnStatus fields for a resource,
+and presents that state to the UI.
+
+```go
+type UIResourceStatus struct {
+  Enabled UIEnableState
+}
+
+type (
+  UIEnableStateOn UIEnableState = iota
+  UIEnableStatePartial
+  UIEnableStateOff
+)
+```
+
+In this model, there's NO source of truth in the model of whether a resource is
+enabled or disabled.  It's only triggered with a toggle.  The source of truth is
+internal to the controller. But interfaces can still get the "result" of being
+disabled from the UIResource aggregator.
+
+The individual controllers get to decide how to process those toggles. If you
+click the button twice before the controller responds, the controller gets to
+decide if that counts as one toggle or two toggles.
+
+NB: This might be a good thing! For resources where it takes a long time to
+start or stop a service, you probably WANT the controller to make smart
+decisions about "queued up" toggles.
```

Comment:
A variant that'd mitigate toggle weirdness:
Create a `UIBoolInput` type, have the UI attach a UIBoolInput value to its enable/disable buttons, and have the reconcilers respond appropriately.

Reply:
ya, that's a good point! let me think on that :thinking:

Reply:
if we did have some sort of a `UIBoolInput` object, do you have something in mind on how the states would combine with multiple inputs?

alternatively - i wonder if we should split up the input method from the current state. So you could pull in Disabled state from a ConfigMap, and have a UIBoolIInput that simply updates the ConfigMap

Reply:
> if we did have some sort of a UIBoolInput object, do you have something in mind on how the states would combine with multiple inputs?

I'm not certain if that's asking about one button with multiple UIBoolInputs or multiple buttons that might each have their own UIBoolInput, so to cover bases:
For the former, we can have some way to determine the name of the input that controls the disabling (presumably something like "DisableOn checks the UIBoolInput named 'disabled'" or "DisableOn requires there to be exactly one input on the button. It must be a UIBoolInput and will control how DisableOn reacts" or "DisableOn takes a button name and a UIBoolInput name")
For the latter, when DisableOn is triggered by a UIButton, it takes the value from that button and ignores all others.

> alternatively - i wonder if we should split up the input method from the current state. So you could pull in Disabled state from a ConfigMap, and have a UIBoolInput that simply updates the ConfigMap

Maybe obvious, but this seems like a continuation of "does Cmd point at FileWatch or does FileWatch point at Cmd?". Using an intermediary like ConfigMap so that both ends just talk to ConfigMaps could clean up a lot of that awkwardness!